### PR TITLE
 Support Boost.MP11 when feasible

### PR DESCRIPTION
**Pull request by CromwellEnage**
* Add are_tagged_arguments_mp11 and is_argument_pack_mp11 metafunctions when Boost.MP11 is usable.
* Predicate requirements can be encoded as Boost.MP11-style quoted metafunctions as well as by MPL binary metafunction classes.
* Argument packs qualify as Boost.MP11-style lists as well as MPL sequences.
* Internal components and test programs use Boost.MP11 and C++11 type traits vice MPL and Boost.TypeTraits when Boost.MP11 is usable.

### CODE REVIEW

**Comment by pdimov** [comment]
You are not supposed to specialize mp11::detail types, sorry.

**Comment by CromwellEnage** [comment]
Oh.  Are there viable alternatives, or do I just need to revert the portion of the changes that allow the ArgumentPack models to also be MP11 lists?

**Comment by pdimov** [comment]
What is, and what isn't, an mp11 list is well defined and not subject to specializations. `empty_arg_list`, for example, isn't. `arg_list<TaggedArg,Next,EmitsErrors>` is a list of exactly size 3, with elements exactly `TaggedArg`, `Next`, and `EmitsErrors`.

**Comment by CromwellEnage** [comment]
I'm still looking for a workaround.  I'm currently stumped at conceptually being unable to do:

``template <typename ...KeyTaggedArgTuples>``
``struct arg_list;``

``template <>``
``struct arg_list<> : public empty_arg_list {};``

``template <typename Key, typename TaggedArg, typename ...Rest>``
``struct arg_list< mp11::mp_list<Key,TaggedArg>, Rest...> : public arg_list<Rest...> {};``

The key problem lies in inheritance, of course, so it'll take me a while to find a way to not rely on it.

**Comment by pdimov** [comment]
If you have `arg_list<KeyTagged...>`, you don't need inheritance or specialization for the zero-arg case. Assuming that the problem is that you want to do lookup by `Key` and that the inner tuples are indeed `mp_list<Key, TaggedArg>` (or `pair`, or `tuple`, or `arg_tuple`, it doesn't matter), you can then use `mp_map_find` for the lookup.

**Comment by pdimov** [comment]
That's assuming that `arg_list` only stores type information. If it has data members, things will be trickier.

File: include/boost/parameter/are_tagged_arguments.hpp [hunk 1]
```diff
@@ -43,6 +43,20 @@ namespace boost { namespace parameter {
     };
 }} // namespace boost::parameter
 
+#if defined(BOOST_PARAMETER_CAN_USE_MP11)
+#include <boost/mp11/integral.hpp>
+
+namespace boost { namespace parameter {
+
+    template <typename TaggedArg0, typename ...TaggedArgs>
+    using are_tagged_arguments_mp11 = ::boost::mp11::mp_bool<
```
Comment (pdimov) [inline include/boost/parameter/are_tagged_arguments.hpp]: This shouldn't be needed. The original is usable as-is.

**Comment by pdimov** [comment]
It does, so you'll need to inherit `arg_list<KeyTagged...>` from `KeyTagged...` and then have

```
template<class Key, class TaggedArg> arg_tuple<Key, TaggedArg>& get( arg_tuple<Key, TaggedArg>& x ) { return x; }
```

and `get<Key>(list)` will get you the correct base (which will contain the `TaggedArg` as a data member.)

This is similar to how `std::tuple` is implemented, except there indices are used for the keys. An example tuple implementation is here:

https://github.com/boostorg/beast/blob/develop/include/boost/beast/core/detail/tuple.hpp

except it's more complex than we need here because it needs to generate an index sequence, whereas we already have the list of keyed bases, so we can just inherit from it directly.

**Comment by pdimov** [comment]
Something like https://godbolt.org/z/JVywuv

File: include/boost/parameter/are_tagged_arguments.hpp (hunk 1 above)
Comment (CromwellEnage) [reply include/boost/parameter/are_tagged_arguments.hpp]: I'll remove that and is_argument_pack_mp11 on the next pull request I submit.

Comment (CromwellEnage) [reply include/boost/parameter/are_tagged_arguments.hpp]: Okay, they're removed.

**Comment by pdimov** [comment]
`operator[]` is trivial to implement, see https://godbolt.org/z/3Qids9

**Pull request merged**